Description: Make build without architecture-specific tuning
Author: Daniel Borkmann <daniel@netsniff-ng.org>
Last-Update: 2012-04-01
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -16,15 +16,13 @@ include(CheckHwTimestamp)
 include(Pod2Man)
 
 add_definitions(
-  -O3
+  -O2
   -fstack-protector
   -fpie
   -std=gnu99
   -fno-strict-aliasing
   -D_FORTIFY_SOURCE=2
   -D_REENTRANT
-  -march=native
-  -mtune=native
   -Wall
   -Wundef
   -Wstrict-prototypes
