Description: Man page path fix
Author: Emmanuel Roullit <emmanuel@netsniff-ng.org>
Last-Update: 2012-04-01
--- a/src/cmake/modules/Pod2Man.cmake
+++ b/src/cmake/modules/Pod2Man.cmake
@@ -47,7 +47,7 @@
 
 	INSTALL(
 		FILES ${CMAKE_CURRENT_BINARY_DIR}/${MANFILE}.${SECTION}.gz
-		DESTINATION share/man/man${SECTION}
+		DESTINATION /usr/share/man/man${SECTION}
     	)
 ENDMACRO(POD2MAN PODFILE MANFILE SECTION)
 
